executing/executing.py

```diff
@@ -509,9 +530,556 @@ def visit_ClassDef(self, node):
+            "@": getattr(ast, "MatMult", ()),
+            "//": ast.FloorDiv,
+            "/": ast.Div,
+            "%": ast.Mod,
+            "+": ast.Add,
+            "-": ast.Sub,
+            "<<": ast.LShift,
+            ">>": ast.RShift,
+            "&": ast.BitAnd,
+            "^": ast.BitXor,
+            "|": ast.BitOr,
+        }
+
+        def except_cleanup(inst, node):
+
+            if inst.opname not in (
+                "STORE_NAME",
+                "STORE_FAST",
+                "STORE_DEREF",
+                "DELETE_NAME",
+                "DELETE_FAST",
+            ):
+                return False
+
+            # This bytecode does something exception cleanup related.
+            # The position of the instruciton seems to be something in the last ast-node of the ExceptHandler
+            # this could be a bug, but it might not be observable in normal python code.
+
+            # example:
+            # except Exception as exc:
+            #     enum_member._value_ = value
+
+            # other example:
+            # STORE_FAST of e was mapped to Constant(value=False)
+            # except OSError as e:
+            #     if not _ignore_error(e):
+            #         raise
+            #     return False
+
+            # STORE_FAST of msg was mapped to print(...)
+            #  except TypeError as msg:
+            #      print("Sorry:", msg, file=file)
+
+            x = node
+            while hasattr(x, "parent"):
+                x = x.parent
+                if isinstance(x, ast.ExceptHandler):
+                    if (x.name or "exc") == inst.argval:
+                        return True
+
+            return False
+
+        def inst_match(opname, **args):
+            """
+            match instruction
+
+            Parameters:
+                opname (string|Seq[stirng]): inst.opname has to be equal or in `opname`
+                **args: every arg has to match inst.arg
+
+            Returns:
+                True if all conditions match the instruction
+
+            """
+            if isinstance(opname, tuple):
+                if instruction.opname not in opname:
+                    return False
+            else:
+                if instruction.opname != opname:
+                    return False
+            return all(getattr(instruction, k) == v for k, v in args.items())
+
+        def node_match(node_type, **args):
+            """
+            match the ast-node
+
+            Parameters:
+                node_type: type of the node
+                **args: every `arg` has to be equal `node.arg`
+                        or `node.arg` has to be an instance of `arg` if it is a type.
+            """
+            return isinstance(node, node_type) and all(
+                isinstance(getattr(node, k), v)
+                if isinstance(v, type)
+                else getattr(node, k) == v
+                for k, v in args.items()
+            )
+
+        def mangled_name(node, name=None):
```

Comment:
I didn't know name mangling applied to names other than attributes, that's really surprising. Honestly in the case of imports it seems like a bug. `import __foo` works, but `class A: import __foo` gives `ModuleNotFoundError: No module named '_A__foo'`. Do you know if that's intentional?

I think this function should:

1. Be moved outside of the class, it's useful more generally
2. Be tested in isolation, something like:

```python
tree = ast.parse("""...""")
assert {mangled_name(n) for n in ast.walk(tree)} == {"foo", "_A__bar", ...}
```

(that's a simplification, it should still raise an exception for nodes that don't have names)

3. Completely replace `attr_names_match` throughout the code.
4. Automatically extract the name from import aliases, so no optional `name` argument is needed.

Reply:
the examples I fond there surprised me also.

here is the real world example (I can not tell you so easily the project because I only saved the file):
```
    def test_mangling(self):
        class A:
            def f():
                __mangled = 1
                __not_mangled__ = 2
                import __mangled_mod
                import __package__.module

        self.assertIn("_A__mangled", A.f.__code__.co_varnames)
        self.assertIn("__not_mangled__", A.f.__code__.co_varnames)
        self.assertIn("_A__mangled_mod", A.f.__code__.co_varnames)
        self.assertIn("__package__", A.f.__code__.co_varnames)
```

to point:
3. I don't want to change the behaviour of the 'SentinelNodeFinder' in this merge-request. Can it be done later?
4. Yes, the reason there is maybe just that I found this case also wired and thought that there might be more of it.

Reply:
OK, only replace the uses of attr_names_match in the new code.